Notify client of apple pay sheet showing

MKevinn opened the pull request:
#### 6ff751a465e39d9f969fd9da863971fb8c2841eb
<pre>
Notify client of apple pay sheet showing
<a href="https://bugs.webkit.org/show_bug.cgi?id=303446">https://bugs.webkit.org/show_bug.cgi?id=303446</a>
<a href="https://rdar.apple.com/163838660">rdar://163838660</a>

Reviewed by Abrar Rahman Protyasha.

Add notification to UIDelegate when Apple Pay payment sheet is about to be presented, allowing
visionOS clients to prepare their UI state before the modal appears. This follows the same
pattern already established for WebAuth modal presentation.

Before presentInScene(), which displays the sheet, is invoked in WebPaymentCoordinatorProxyIOS,
`notifyWillPresentPaymentUI` is called to send notification that the sheet will appear. The
method is called via IPC from NetworkProcess to UIProcess, following the same pattern as other
payment coordinator communication.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::notifyWillPresentPaymentUI):
Implement by sending IPC message to NetworkProcessProxy in UIProcess.

* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
Invoke notifyWillPresentPaymentUI immediately before presentInScene() call.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
Add NotifyWillPresentPaymentUI IPC message.

* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::notifyWillPresentPaymentUI):
Implement notification by calling the existing _webViewWillPresentModalUI: UIDelegate method

* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::Internals::notifyWillPresentPaymentUI):

Canonical link: <a href="https://commits.webkit.org/303902@main">https://commits.webkit.org/303902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5361b984b6861eb2886d0a2cf7e4dcacccf645f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141456 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6254 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102414 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136826 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120025 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83213 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4757 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2375 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113914 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144102 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6060 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110782 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6142 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110985 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28160 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4609 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116280 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59807 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6112 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34553 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5958 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69576 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6203 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
```diff
@@ -150,6 +154,10 @@
 
+#if PLATFORM(VISION)
+    page->uiClient().willPresentModalUI(*page);
+#endif
```
The platform guard is unnecessary, given `ENABLE_APPLE_PAY_REMOTE_UI_USES_SCENE` is only true for visionOS

aprotyas commented on Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
```diff
@@ -71,14 +71,16 @@
             if (!paymentCoordinatorProxy->m_authorizationPresenter)
                 return completionHandler(false);
 
+            paymentCoordinatorProxy->checkedClient()->notifyWillPresentPaymentUI(webPageProxyID);
+
             paymentCoordinatorProxy->m_authorizationPresenter->presentInScene(sceneIdentifier, bundleIdentifier, WTFMove(completionHandler));
```
`notifyWillPresentPaymentUI` is an async IPC message, so we no longer have a guarantee on the call ordering between `willPresentModalUI()` and `presentInScene()`. I think it'll be wrong for us to present and then notify the client, right?

You probably want to make `notifyWillPresentPaymentUI` asynchronously call a completion handler after it notifies the client.

i.e.

```cpp
void NetworkConnectionToWebProcess::notifyWillPresentPaymentUI(WebPageProxyIdentifier webPageProxyIdentifier, CompletionHandler<void(void)>&& completionHandler)
{
    networkProcess().parentProcessConnection()->sendWithAsyncReply(Messages::NetworkProcessProxy::NotifyWillPresentPaymentUI(webPageProxyIdentifier), WTFMove(completionHandler));
}
```

You would then change the call site like this:

```cpp
  paymentCoordinatorProxy->checkedClient()->notifyWillPresentPaymentUI(webPageProxyID, [weakThis = WTFMove(weakThis), sceneIdentifier, bundleIdentifier, completionHandler = WTFMove(completionHandler)] mutable {
      RefPtr paymentCoordinatorProxy = weakThis.get();
      if (!paymentCoordinatorProxy)
          return completionHandler(false);

      if (!paymentCoordinatorProxy->m_authorizationPresenter)
          return completionHandler(false);

      paymentCoordinatorProxy->m_authorizationPresenter->presentInScene(sceneIdentifier, bundleIdentifier, WTFMove(completionHandler));
  });
```

MKevinn replied:
Resolved offline. We don't want to delay showing Apple Pay because of this.

aprotyas reviewed (commented):
One comment about the async nature of the payment coordinator IPC.

aprotyas commented on Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
```diff
@@ -152,6 +156,15 @@
+
+void NetworkProcessProxy::notifyWillPresentPaymentUI(WebPageProxyIdentifier webPageProxyIdentifier)
+{
+    auto page = WebProcessProxy::webPage(webPageProxyIdentifier);
```
in adherence with our safer-cpp guidance (we are explicit about refcounted smart pointer types)
```suggestion
    RefPtr page = WebProcessProxy::webPage(webPageProxyIdentifier);
```

Commit:
Notify client of apple pay sheet showing
https://bugs.webkit.org/show_bug.cgi?id=303446
rdar://163838660

Reviewed by Abrar Rahman Protyasha.

Add notification to UIDelegate when Apple Pay payment sheet is about to be presented, allowing
visionOS clients to prepare their UI state before the modal appears. This follows the same
pattern already established for WebAuth modal presentation.

Before presentInScene(), which displays the sheet, is invoked in WebPaymentCoordinatorProxyIOS,
`notifyWillPresentPaymentUI` is called to send notification that the sheet will appear. The
method is called via IPC from NetworkProcess to UIProcess, following the same pattern as other
payment coordinator communication.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::notifyWillPresentPaymentUI):
Implement by sending IPC message to NetworkProcessProxy in UIProcess.

* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
Invoke notifyWillPresentPaymentUI immediately before presentInScene() call.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
Add NotifyWillPresentPaymentUI IPC message.

* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::notifyWillPresentPaymentUI):
Implement notification by calling the existing _webViewWillPresentModalUI: UIDelegate method

* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::Internals::notifyWillPresentPaymentUI):

Canonical link: https://commits.webkit.org/303902@main
